File: lintcode/permutations.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/permutations/

class Solution:
    def helper(self, nums, visited, path, ans):
        if len(path) == len(nums):
            ans.append(path[:])
            return

        for i in xrange(len(nums)):
            if nums[i] in visited:
                continue
            visited.add(nums[i])
            path.append(nums[i])
            self.helper(nums, visited, path, ans)
            visited.remove(nums[i])
            path.pop()

    """
    @param nums: A list of Integers.
    @return: A list of permutations.
    """
    def permute(self, nums):
        # write your code here
        if not nums:
            return []

        ans = []
        self.helper(nums, set(), [], ans)
        return ans

if __name__ == '__main__':
    s = Solution()
    print s.permute([1,2,3])

File: leetcode-2/permutations.cc
// Question: https://leetcode.com/problems/permutations/

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        unordered_set<int> visited;
        vector<int> path;
        vector<vector<int>> ans;
        helper(path, ans, visited, nums);
        return ans;
    }

    void helper(vector<int>& path, vector<vector<int>>& ans, unordered_set<int>& visited, vector<int>& nums) {
        if (path.size() == nums.size()) {
            ans.push_back(path);
        } else if (path.size() < nums.size()) {
            for (int i = 0; i < nums.size(); ++i) {
                int n = nums[i];
                if (visited.count(n) == 0) {
                    path.push_back(n);
                    visited.insert(n);
                    helper(path, ans, visited, nums);
                    visited.erase(n);
                    path.pop_back();
                }
            }
        }
    }
};
